src/backend/distributed/commands/table.c

```diff
@@ -898,7 +945,33 @@ PreprocessAlterTableAddIndexConstraint(AlterTableStmt *alterTableStatement, Oid
 	ObjectAddressSet(ddlJob->targetObjectAddress, RelationRelationId, relationId);
 	ddlJob->startNewTransaction = false;
 	ddlJob->metadataSyncCommand = ddlCommand;
-	ddlJob->taskList = DDLTaskList(relationId, ddlCommand);
+
+
+	if (constraint->contype == CONSTR_FOREIGN)
+	{
+		Oid rightRelationId = RangeVarGetRelid(constraint->pktable, NoLock,
+											   false);
+
+		if (IsCitusTableType(rightRelationId, REFERENCE_TABLE))
+		{
+			SetLocalMultiShardModifyModeToSequential();
+		}
+
+		bool referencedIsLocalTable = !IsCitusTable(rightRelationId);
+		if (referencedIsLocalTable)
+		{
+			ddlJob->taskList = NIL;
```

Comment:
what happens in this case?

Reply:
This block is to mimick the behaviour in preprocess function for foreign keys that are not deparsed. I copied it from https://github.com/citusdata/citus/blob/28ed013a91a7a3b3048c716153856580294571f4/src/backend/distributed/commands/table.c#:~:text=bool%20referencedIsLocalTable%20%3D%20!,%7D

I believe I do not have a regress test for this case yet. It looks like when there is no distributed table involved in the constraint this code path is hit.

But then if we do not have to propagate the command to workers, we should be fine with the original command. Now I think we do not need it in this PR, right?

Reply:
Looking at this again, PreprocessAlterTableStmt function returns empty tasklist for cases citus will error out later on. For example if a relation in a FOREIGN KEY constraint is not a citus table, it returns empty tasklist. To conform to this contract, I do the check on right relation in this new function as it returns early. 

I added a test case covering this code as well.

Reply:
thanks for adding the comment